dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/registry/MasterRegistryClient.java

```diff
@@ -357,9 +299,99 @@ private void failoverMaster(String masterHost) {
         logger.info("master failover end");
     }
 
-    public InterProcessMutex blockAcquireMutex() throws Exception {
-        InterProcessMutex mutex = new InterProcessMutex(getZkClient(), getMasterLockPath());
-        mutex.acquire();
-        return mutex;
+    public void blockAcquireMutex() {
+        registryClient.getLock(registryClient.getMasterLockPath());
     }
+
+    public void releaseLock() {
+        registryClient.releaseLock(registryClient.getMasterLockPath());
+    }
+
+    /**
+     * master config
+     */
+    @Autowired
+    private MasterConfig masterConfig;
+
+    /**
+     * heartbeat executor
+     */
+    private ScheduledExecutorService heartBeatExecutor;
+
+    /**
+     * master start time
+     */
+    private String startTime;
+
+    private String localNodePath;
+
+    @PostConstruct
+    public void init() {
+        this.startTime = DateUtils.dateToString(new Date());
+        this.heartBeatExecutor = Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory("HeartBeatExecutor"));
+        registryClient.init();
+    }
+
+    /**
+     * registry
+     */
+    public void registry() {
+        String address = NetUtils.getAddr(masterConfig.getListenPort());
+        localNodePath = getMasterPath();
+        registryClient.persistEphemeral(localNodePath, "");
+        registryClient.addConnectionStateListener(new MasterRegistryConnectStateListener());
+        int masterHeartbeatInterval = masterConfig.getMasterHeartbeatInterval();
+        HeartBeatTask heartBeatTask = new HeartBeatTask(startTime,
+                masterConfig.getMasterMaxCpuloadAvg(),
+                masterConfig.getMasterReservedMemory(),
+                Sets.newHashSet(getMasterPath()),
+                Constants.MASTER_TYPE,
+                registryClient);
+
+        this.heartBeatExecutor.scheduleAtFixedRate(heartBeatTask, masterHeartbeatInterval, masterHeartbeatInterval, TimeUnit.SECONDS);
+        logger.info("master node : {} registry to ZK successfully with heartBeatInterval : {}s", address, masterHeartbeatInterval);
+
+    }
+
+    class MasterRegistryConnectStateListener implements RegistryConnectListener {
+
+        @Override
+        public void notify(RegistryConnectState newState) {
+            if (RegistryConnectState.RECONNECTED == newState) {
+                registryClient.persistEphemeral(localNodePath, "");
```

Comment:
When reconnect to zookeeper, if the session has not expired, the previous ephemeral node may still exist.
If we delete the previous node, will cause the current instance to be killed.
```
public void persistEphemeral(String key, String value) {
        try {
            if (isExisted(key)) {
                client.delete().deletingChildrenIfNeeded().forPath(key);
            }
            client.create().creatingParentsIfNeeded().withMode(CreateMode.EPHEMERAL).forPath(key, value.getBytes(StandardCharsets.UTF_8));
        } catch (Exception e) {
            throw new RegistryException("zookeeper persist ephemeral error", e);
        }
    }
```